engines/dfc_provider/app/services/dfc_product_type_factory.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'singleton'
+
+class DfcProductTypeFactory
+  include Singleton
+
+  def self.for(dfc_id)
+    instance.for(dfc_id)
+  end
+
+  def initialize
+    @product_types = {}
+
+    populate_product_types
+  end
+
+  def for(dfc_id)
+    @product_types[dfc_id]
+  end
+
+  private
+
+  def populate_product_types
+    DfcLoader.connector.PRODUCT_TYPES.topConcepts.each do |product_type|
+      stack = []
+      record_type(stack, product_type.to_s)
+    end
+  end
+
+  def record_type(stack, product_type)
+    name = product_type.to_s
+    current_stack = stack.dup.push(name)
+
+    type = call_dfc_product_type(current_stack)
+
+    id = type.semanticId
+    @product_types[id] = type
```

Comment:
What's the difference between `type` and `product_type`? I was just wondering if we can get rid of `call_dfc_product_type` altogether.

Reply:
`type` is the actual object representing a product type, and `product_type` is the name of the method to call to get the actual object. 
I had another look and you can indeed get rid of `call_dfc_product_type` see : [de95b0c](https://github.com/openfoodfoundation/openfoodnetwork/pull/11817/commits/de95b0c7baf636da7dceaa380d671fbe5e490d06)

I am not sure why I though we needed to record the path, I think I just spent to long working on it. It's nice to get an external perspective :pray: